### PR TITLE
Address issues #3 and #8

### LINKED ISSUE #3: Impact of INT on Path MTU
As transit INT devices expand the size of data packets by adding INT metadata to the packets, end hosts have to be able to discover path MTU accounting for INT. This aspect is not well addressed in the current specification. 

A desirable way of handling this would be for a INT transit hop at which packet size after adding INT metadata exceeds outgoing link MTU, to send an ICMP Destination Unreachable message with code = “fragmentation needed and DF set”, and an appropriately calculated MTU reported in the ICMP message, allowing end hosts to discover the right path MTU with INT being performed in the network.
 
The path MTU aspect needs to be clarified and expanded upon in the dataplane specification.

### LINKED ISSUE #8: Reserved INT instruction bits: clarify behavior for forward compatibility
There are currently 8 reserved INT instruction bits. We need to specify required behavior when receiving a packet with one or more of the reserved INT instruction bits set. The typical reserved field behavior, setting to zero on transmission and ignoring received values, is not sufficient since any future use of these bits would break backwards compatibility. If software versions are mixed in the same network, nodes running the older version will add less INT metadata than nodes running the newer version. This will make it difficult for the monitor to determine which INT metadata came from which node. It is also likely that there will be a mismatch between the amount of INT metadata added and the value of the instruction count, leading to an incorrect increment of the length in the IP header, and UDP header if present.

In order to allow the reserved bits to be used in future versions with backward compatibility, we need to check whether any of the reserved INT instruction bits is set. At first glance there seem to be two possible approaches:
1. A node receiving a packet with one or more of the reserved INT instruction bits set must not add any metadata to that INT hop-by-hop header.
2. A node receiving a packet with one or more of the reserved INT instruction bits set must add the corresponding metadata filled with "0xFF" in each octet.

Any node adding an INT hop-by-hop header must set all reserved instruction bits to 0.

### PR DESCRIPTION
**Pull request by mhira1**
Edits to INT.mdk to elaborate on on MTU handling (fixes #3) and handling of receiving reserved bits set at a transit hop (fixes #8)

### CODE REVIEW

File: telemetry/specs/INT.mdk [hunk 1]
```diff
@@ -298,18 +298,64 @@ exhausting header space in the case of a forwarding loop or any other anomalies,
...
+Alternatively, each INT hop may report an MTU only accounting for the metadata
+it inserts. This would enable the path MTU discovery source converge to a
+precise path MTU, at the cost of receiving more ICMP messages, one from each
+INT hop.
+
+Regardless of whether or not an INT transit switch participates in Path MTU
+discovery, if it cannot insert all requested metadata because doing so will
+cause the packet length to exceed egress link MTU, it
+**must not insert any metadata and set the M bit** in the
+INT header, indicating that egress MTU was exceeded at an INT hop.
```
Comment (jklr) [inline telemetry/specs/INT.mdk:334]: Though we say below at the M bit description, let's clarify here that the device must not add any metadata to the packet.

File: telemetry/specs/INT.mdk [hunk 2]
```diff
@@ -833,7 +905,7 @@ INT Shim Header for VXLAN-GPE, destination INT type:
    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 `
 
-INT Metadata Header and Metadata Stack, 
+INT Metadata Header and Metadata Stack,
 followed by encapsulated Ethernet payload:
 `
     0                   1                   2                   3
```
Comment (jklr) [inline telemetry/specs/INT.mdk:913]: Please revise the header examples with the newly added M bit.

File: telemetry/specs/INT.mdk [hunk 3]
```diff
@@ -983,30 +1055,30 @@ header int_header_t {
     bit<4>  instruction_mask_0811;
     bit<4>  instruction_mask_1215;
     bit<16> rsvd2;
-}   
-    
+}
```
Comment (jklr) [inline telemetry/specs/INT.mdk:1061]: M bit in the int_header_t definition.

File: telemetry/specs/INT.mdk [hunk 4]
```diff
@@ -1380,7 +1452,7 @@ control Int_egress(inout headers hdr, inout metadata meta,
 
     apply{
         if(hdr.int_header.isValid()) {
-            if(hdr.int_header.max_hop_cnt != hdr.int_header.total_hop_cnt 
+            if(hdr.int_header.max_hop_cnt != hdr.int_header.total_hop_cnt
```
Comment (jklr) [inline telemetry/specs/INT.mdk:1459]: We can add a condition that checks MTU limit and an action that sets the M bit.

File: telemetry/specs/INT.mdk [hunk 5]
```diff
@@ -298,18 +298,64 @@ exhausting header space in the case of a forwarding loop or any other anomalies,
...
+(Instruction Count\*4\*INT Hop Count + 8), based on conservative values of
+total number of INT hops and instruction count being used, will prevent egress
+MTU being exceeded due to INT metadata insertion at INT hops.
+
+* An INT source/transit switch may optionally aid in dynamic discovery of Path MTU
+along flows being monitored by INT by transmitting ICMP message as per Path MTU
+Discovery mechanisms of the corresponding IP protocol
+(RFC 1191 for IPv4, RFC 1981 for IPv6). An INT source or transit switch may report
+a conservative MTU in the ICMP message, accounting for metadata insertion at
+downstream INT hops and assuming uniform egress MTU at downstream hops. This
```
Comment (mickeyspiegel) [inline telemetry/specs/INT.mdk]: Not just downstream hops. All hops including upstream and downstream should be subtracted from the underlying real MTU of the egress link.

File: telemetry/specs/INT.mdk (hunk 1 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk:334]: It is already clarified both here and below in the M bit description.

File: telemetry/specs/INT.mdk (hunk 2 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk:913]: Done.

File: telemetry/specs/INT.mdk (hunk 3 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk:1061]: Done.

File: telemetry/specs/INT.mdk (hunk 4 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk:1459]: Int_egress currently does not have access to egress MTU. Can you please modify the P4 program to pass in egress MTU to Int_egress and set the M bit if needed in Int_egress?

File: telemetry/specs/INT.mdk [hunk 6]
```diff
@@ -298,18 +298,64 @@ exhausting header space in the case of a forwarding loop or any other anomalies,
...
+
+An INT source needs to be able to insert 8 bytes of INT header plus Instruction Count \* 4
+bytes of its own metadata. If inserting 8 bytes of INT header itself causes
+egress link MTU to be exceeded, INT will not be initiated for such
+packets (**TBD - handling of such scenarios**). If there is enough room to
+insert INT header, but not first-hop INT metadata at the source, the source must
+initiate INT and set the M bit at the first hop itself.
+
+In theory, an INT transit switch can perform IPv4 fragmentation to overcome
+egress MTU limitation when inserting its metadata. However, IPv4 fragementation
```
Comment (mickeyspiegel) [inline telemetry/specs/INT.mdk]: typo: fragementation --> fragmentation

File: telemetry/specs/INT.mdk [hunk 7]
```diff
@@ -298,18 +298,64 @@ exhausting header space in the case of a forwarding loop or any other anomalies,
...
+can have adverse impact on applications. Moreover, IPv6 packets cannot be
+fragmented at intermediate hops. Also, fragmenting packets at INT transit hops,
+with or without copying preceding INT metadata into fragments imposes
+extra complexity of corelating fragments in the INT monitoring engine.
```
Comment (mickeyspiegel) [inline telemetry/specs/INT.mdk]: typo: corelating --> correlating

**Comment by mhira1** [comment]
Incorporated all comments, except for handling egress MTU and M bit in the P4 program. @jklr can you please make this change to the P4 program?

File: telemetry/specs/INT.mdk [hunk 8]
```diff
@@ -612,14 +649,49 @@ metadata to the packet. If Total Hop Count (see below) equals Max Hop Count, a
...
+header.
+  - When adding a new metadata, each device **must prepend** its metadata in
+front of the metadata that are already added by the upstream devices. This is
+similar to the push operation on a stack. Hence, the most recently added
+metadata appears at the top of the stack. The device must add metadata in the
+order of bits set in the instruction bitmap.
+  - If a device is unable to provide a metadata value specified in the
+instruction bitmap because its value is not available, it must add a special
+reserved value  0xFFFF_FFFF indicating “invalid”.
+  - If a device cannot add all the number of bytes (i.e., Instruction Count\*4) required by the instruction bitmap (irrespective of the availability of the
```
Comment (mickeyspiegel) [inline telemetry/specs/INT.mdk]: Break the long line in two?

**Review by mickeyspiegel — APPROVED** [review]

File: telemetry/specs/INT.mdk (hunk 5 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk]: Agreed, fixed this.

File: telemetry/specs/INT.mdk (hunk 6 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk]: Done

File: telemetry/specs/INT.mdk (hunk 7 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk]: Done

File: telemetry/specs/INT.mdk (hunk 8 above)
Comment (mhira1) [reply telemetry/specs/INT.mdk]: Done

**Pull request merged**